### ISSUE TITLE
Setting winhighlight on a floating window leaks into quickfix windows

### ISSUE THREAD

**Issue by b0o** [neovim/neovim#18283]
### Neovim version (nvim -v)

v0.8.0-dev+9-g1facad234

### Vim (not Nvim) behaves the same?

n/a

### Operating system/version

Linux

### Terminal name/version

Happens in both alacritty (latest) and kitty (latest)

### $TERM environment variable

tmux-256color, xterm-kitty

### Installation

AUR: neovim-git

### How to reproduce the issue

1. `nvim --clean`
2. `:lua vim.api.nvim_win_set_option(vim.api.nvim_open_win(vim.api.nvim_create_buf(false, true), false, { relative = "editor", width = 1, height = 1, row = 1, col = 1 }), "winhighlight", "Normal:Search")`
3. `:copen`

### Expected behavior

![2022-04-27_02-27-52_Alacritty](https://user-images.githubusercontent.com/21299126/165487729-be1776e2-6b6e-456f-8302-b5a90d5a4f05.png)

Focus quickfix window, `:echo &winhighlight` should be blank

### Actual behavior

![2022-04-27_02-27-03_Alacritty](https://user-images.githubusercontent.com/21299126/165487574-e0016132-e2b1-4ea7-8874-0ab5040b00d6.png)

Focus quickfix window, `:echo &winhighlight` is `Normal:Search`

Also happens with loclist window.

**Comment by zeertzjq** [comment]
Is this really a bug?

**Comment by b0o** [comment]
I believe so. `winhighlight` is supposed to be window-local. Why should it affect any window other than the one it is set on?

In case it's not clear from my repro, I am openeing a new floating window and immediately setting winhighlight on it. Then, in a separate command, I am opening the quickfix window (which should have nothing to do with the floating window).

**Comment by bfredl** [comment]
Right, it should copy `winhl` from the current window, which still is the above split, if I understand correctly? then it is a bug indeed.

**Comment by b0o** [comment]
@bfredl Should `winhl` be copied at all? Doing this, that is not the behavior I see:

1. `nvim --clean`
2. `:setlocal winhl=Normal:Search`
3. `:copen`

**Comment by zeertzjq** [comment]
Oh, I didn't notice that the floating window is not focused. Then this is a bug.

**Comment by bfredl** [comment]
> My assumption, though, is that nvim_win_set_option behaves like setlocal.

Yes, this is the intention, but there is a bug in the implementation apparently (not the first time..)

**Comment by bfredl** [comment]
oops

**Comment by zeertzjq** [comment]
This is because the quickfix window is split from the last window in the window list:
```c
  if (IS_QF_STACK(qi) && cmdmod.split == 0) {
    // Create the new quickfix window at the very bottom, except when
    // :belowright or :aboveleft is used.
    win_goto(lastwin);
  }
  // Default is to open the window below the current window
  if (cmdmod.split == 0) {
    flags = WSP_BELOW;
  }
  flags |= WSP_NEWLOC;
  if (win_split(height, flags) == FAIL) {
    return FAIL;  // not enough room for window
  }
```

It may be considered a Vim bug as well.

**Comment by b0o** [comment]
I would expect these two sets of steps to yield the same result, but they don't. Might be relevant?

#### setlocal winhighlight

1. `nvim --clean`
2. `:setlocal winhighlight=Normal:Search`
3. `:copen`

![2022-04-27_02-58-26_Alacritty](https://user-images.githubusercontent.com/21299126/165493393-d180c067-c195-47e0-b2ed-a6232d4c63ca.png)

#### vim.wo.winhighlight

1. `nvim --clean`
2. `:lua vim.wo.winhighlight = 'Normal:Search'`
3. `:copen`

![2022-04-27_02-59-27_Alacritty](https://user-images.githubusercontent.com/21299126/165493551-9f87f9ab-5e53-4b72-a3fe-dc2ad7405b1e.png)

**Comment by bfredl** [comment]
> This is because the quickfix window is split from the last window in the window list.

Hmm at very least it should be changed to last non-floating then.

**Comment by b0o** [comment]
This seems to be the same underlying issue as #14670

**Comment by lewis6991** [comment]
> > My assumption, though, is that nvim_win_set_option behaves like setlocal.
> 
> Yes, this is the intention, but there is a bug in the implementation apparently (not the first time..)

I don't think this is correct `nvim_win_set_option` behaves like `set` for window local options. #9110 introduced this.

Anyway I don't think there is an issue here with `nvim_set_win_option`, it simply works the way it does. The issue here is what zeer pointed out.

**Comment by glepnir** [comment]
Fixed by #33807
